### PR TITLE
Enhancement/permission changer

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
Adds remove/add to permissions

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 1]
```diff
@@ -31,58 +37,72 @@
...
+	@Override
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		for (Entity entity : entities.getAll(event)) {
+			PermissionAttachment perm = getPermission(entity);
+			for (String string : (String[]) delta)
+				perm.setPermission(string, mode == ChangeMode.ADD);
+		}
+	}
+
+	private final String PERMISSION_TAG = "skript-permissions";
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:99]: Can this be static and moved to the top with the other field?

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
Nice PR \:D

Does Permissible work with other permission plugins, in other words does the `getPermission` method support other permission plugins?

**Review by AyhamAl-Ali — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/data/BukkitClasses.java [hunk 2]
```diff
@@ -1837,4 +1837,5 @@ public String toVariableNameString(Attribute a) {
 				})
 				.serializer(new EnumSerializer<>(Attribute.class)));
 	}
+
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:1840]:
```suggestion
```
Can be removed since its unrelated

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 3]
```diff
@@ -31,58 +37,73 @@
...
+@Description("Returns all permissions of the defined permissible(s). A permissible is an object like an entity that can have permissions.")
 @Examples("set {_permissions::*} to all permissions of the player")
-@Since("2.2-dev33")
+@Since("2.2-dev33, INSERT VERSION (Changers)")
 public class ExprPermissions extends SimpleExpression<String> {
-	
+
 	static {
-		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] permissions (from|of) %players%", "[(all [[of] the]|the)] %players%'[s] permissions");
+		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] permissions (from|of) %entities%", "[(all [[of] the]|the)] %entities%'[s] permissions");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:47]:
```suggestion
		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY,
			"[all [[of] the]|the] permissions (from|of) %entities%",
			"[all [of] the] %entities%'[s] permissions"
		);
```
Technically a breaking syntax change but `all player's permissions` makes no sense

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 4]
```diff
@@ -31,58 +37,73 @@
...
-import org.bukkit.permissions.PermissionAttachment;
-import org.bukkit.permissions.PermissionAttachmentInfo;
-import org.eclipse.jdt.annotation.Nullable;
-
-import java.util.HashSet;
-import java.util.Set;
-
 @Name("All Permissions")
-@Description("Returns all permissions of the defined player(s). Note that the modifications to resulting list do not actually change permissions.")
+@Description("Returns all permissions of the defined permissible(s). A permissible is an object like an entity that can have permissions.")
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:41]:
```suggestion
@Description("All permissions of the provided permissible(s). A permissible is anything that can have permissions (like the player).")
```

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 5]
```diff
@@ -31,58 +37,73 @@
...
 	@Override
 	public Class<String> getReturnType() {
 		return String.class;
 	}
-	
+
 	@Override
 	public String toString(@Nullable Event event, boolean debug) {
-		return "permissions " + " of " + players.toString(event, debug);
+		return "permissions " + " of " + entities.toString(event, debug);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:83]:
```suggestion
		return "permissions of " + entities.toString(event, debug);
```

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 6]
```diff
@@ -31,58 +37,73 @@
...
+	}
+
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE)
+			return CollectionUtils.array(String.class, String[].class);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:89]:
```suggestion
			return CollectionUtils.array(String[].class);
```
Only the one is needed I believe

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 7]
```diff
@@ -31,58 +37,73 @@
...
+		return null;
+	}
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		for (Entity entity : entities.getAll(event)) {
+			PermissionAttachment perm = getPermission(entity);
+			for (String string : (String[]) delta)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:97]: Casting delta itself can cause issues sometimes (it has happened with skript-gui before)

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
How does this interact with something like a permissions plugin?

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java (hunk 7 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:97]: I don't see it here.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:100]: It only happens when you spam the expression. This seems like an issue would need to address
```
java.lang.ClassCastException: class [Ljava.lang.Object; cannot be cast to class [Ljava.lang.String; ([Ljava.lang.Object; and [Ljava.lang.String; are in module java.base of loader 'bootstrap')
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.expressions.ExprPermissions.change(ExprPermissions.java:108)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.effects.EffChange.execute(EffChange.java:280)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.Effect.run(Effect.java:50)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:62)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:90)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.sections.EffSecSpawn.lambda$walk$1(EffSecSpawn.java:149)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.addEntity(CraftRegionAccessor.java:533)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.spawn(CraftRegionAccessor.java:512)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.spawn(CraftRegionAccessor.java:506)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.spawn(CraftRegionAccessor.java:497)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.entity.EntityData.spawn(EntityData.java:426)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.sections.EffSecSpawn.walk(EffSecSpawn.java:167)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:90)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.Trigger.execute(Trigger.java:52)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.events.EvtScript.runTrigger(EvtScript.java:88)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.events.EvtScript.register(EvtScript.java:68)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.SkriptEvent.postLoad(SkriptEvent.java:151)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.lambda$loadScripts$9(ScriptLoader.java:573)
[21:32:12 ERROR]: #!#!     at java.base/java.util.ArrayList.removeIf(ArrayList.java:1672)
[21:32:12 ERROR]: #!#!     at java.base/java.util.ArrayList.removeIf(ArrayList.java:1660)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.lambda$loadScripts$10(ScriptLoader.java:569)
[21:32:12 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
[21:32:12 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
[21:32:12 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadScripts(ScriptLoader.java:513)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadScripts(ScriptLoader.java:460)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.SkriptCommand.onCommand(SkriptCommand.java:189)
[21:32:12 ERROR]: #!#!     at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45)
[21:32:12 ERROR]: #!#!     at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:155)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftServer.dispatchCommand(CraftServer.java:916)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftServer.dispatchServerCommand(CraftServer.java:879)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.dedicated.DedicatedServer.bg(DedicatedServer.java:500)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:447)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1397)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.v(MinecraftServer.java:1173)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305)
[21:32:12 ERROR]: #!#!     at java.base/java.lang.Thread.run(Thread.java:833)
```

**Review by Moderocky — APPROVED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprPermissions.sk [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+		# Entities have default bukkit permissions.
+		add "permission.test" to permissions of event-entity
+		assert permissions of event-entity contains "permission.test" with "permission failed to be added"
+		add "permission.test1" and "permission.test2" to permissions of event-entity
+		assert permissions of event-entity contains "permission.test", "permission.test1" and "permission.test2" with "multiple permissions failed to be added"
+		remove "permission.test1" from permissions of event-entity
+		assert permissions of event-entity contains "permission.test" and "permission.test2" with "failed to remove permission"
+		clear permissions of event-entity
+		assert permissions of event-entity doesn't contain "permission.test", "permission.test1" and "permission.test2" with "permissions were not cleared"
+		delete event-entity
```
Comment (APickledWalrus) [inline src/test/skript/tests/syntaxes/expressions/ExprPermissions.sk:12]: does this actually work? I don't think the entity has spawned in yet

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 9]
```diff
@@ -31,58 +40,117 @@
...
 
 	@Override
 	@Nullable
-	protected String[] get(Event e) {
-		final Set<String> permissions = new HashSet<>();
-		for (Player player : players.getArray(e))
-			for (final PermissionAttachmentInfo permission : player.getEffectivePermissions())
-				permissions.add(permission.getPermission());
-		return permissions.toArray(new String[permissions.size()]);
+	protected String[] get(Event event) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:72]: I'm not sure how much it matters, but maybe it'd be a good idea to override the stream method too?

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 10]
```diff
@@ -31,58 +40,117 @@
...
+					if (attachment == null)
+						continue;
+					for (String permission : attachment.getPermissions().keySet()) {
+						if (mode == ChangeMode.DELETE) {
+							attachment.unsetPermission(permission);
+							continue;
+						} else if (permissions.contains(permission)) {
+							attachment.unsetPermission(permission);
+						}
+					}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:111]:
```suggestion
					if (mode == ChangeMode.DELETE) {
						for (String permission : attachment.getPermissions().keySet())
							attachment.unsetPermission(permission);
					} else {
						for (String permission : permissions)
							attachment.unsetPermission(permission);
					}
```
Could even reduce it down to
```suggestion
					for (String permission : mode == ChangeMode.DELETE ? attachment.getPermissions().keySet() : permissions)
						attachment.unsetPermission(permission);
```
(it's a bit long though for the loop line)

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 11]
```diff
@@ -31,58 +40,117 @@
...
+			if (delta == null)
+				return;
+			for (Object permission : delta) {
+				if (permission == null)
+					continue;
+				permissions.add((String) permission);
+			}
+		}
+		if (mode == ChangeMode.DELETE || mode == ChangeMode.REMOVE) {
+			for (Entity entity : entities.getAll(event)) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:99]: Is there a reason you're using getAll over getArray?

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 12]
```diff
@@ -31,58 +40,117 @@
...
+	private PermissionAttachment getPermission(Entity entity) {
+		Skript instance = Skript.getInstance();
+		for (PermissionAttachmentInfo info : entity.getEffectivePermissions()) {
+			PermissionAttachment attachment = info.getAttachment();
+			if (attachment == null)
+				continue;
+			if (attachment.getPlugin().equals(instance))
+				return attachment;
+		}
+		return  entity.addAttachment(instance);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:153]:
```suggestion
		return entity.addAttachment(instance);
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprPermissions.sk (hunk 8 above)
Comment (TheLimeGlass) [reply src/test/skript/tests/syntaxes/expressions/ExprPermissions.sk:12]: Yes it does, I did broadcast debugs, otherwise I wouldn't have known the default permissions Bukkit adds.

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java (hunk 9 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:72]: You mean the iterator method? Overriding the iterator method is what Njol considers to be better for performance, it's main reasoning is for looping, example looping blocks. This doesn't make too much sense as it typically won't be used like that.

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java (hunk 10 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:111]: Attachment is null if it's a default permission. There must be a null check.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:111]: For the first suggest, you need to check if the permission being removed exists in their permission set. You're removing that condition check.

**Comment by sovdeeth** [comment]
I like the addition and the code looks good, but I'm concerned about the changes to permissions being temporary. Is there a way we can make them permanent without a ton of effort? I think users will get confused and wonder why the changes aren't lasting over restarts, no matter how much documentation we have, and with something as critical as permissions, I'm not sure that's something we should just accept.

**Comment by TheLimeGlass** [comment]
> I like the addition and the code looks good, but I'm concerned about the changes to permissions being temporary. Is there a way we can make them permanent without a ton of effort? I think users will get confused and wonder why the changes aren't lasting over restarts, no matter how much documentation we have, and with something as critical as permissions, I'm not sure that's something we should just accept.

No. The whole purpose of permission plugins is to allow for database saving. Bukkit does not have a built permission saving.

Skript has variables which allows the user to save the permissions themselves, removing the need for a permission plugin, as Skript becomes the permission plugin handler.

**Review by Moderocky — APPROVED** [review]
This seems like a good addition -- since we already have Vault as a dependency is there anything we can do on the permission side there (i.e. notifying vault that we want to add a permission to something?)

**Comment by sovdeeth** [comment]
> > I like the addition and the code looks good, but I'm concerned about the changes to permissions being temporary. Is there a way we can make them permanent without a ton of effort? I think users will get confused and wonder why the changes aren't lasting over restarts, no matter how much documentation we have, and with something as critical as permissions, I'm not sure that's something we should just accept.
> 
> No. The whole purpose of permission plugins is to allow for database saving. Bukkit does not have a built permission saving.
> 
> Skript has variables which allows the user to save the permissions themselves, removing the need for a permission plugin, as Skript becomes the permission plugin handler.

Hmm, makes sense. I would really like to see this work with the user's permissions plugin via hooks, if possible, but I'd accept that in a future PR too.

**Comment by TheLimeGlass** [comment]
> 
> Hmm, makes sense. I would really like to see this work with the user's permissions plugin via hooks, if possible, but I'd accept that in a future PR too.

You can manipulate the permissions and the permissions plugin should respect the memory list value.

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 13]
```diff
@@ -18,71 +18,135 @@
...
-public class ExprPermissions extends SimpleExpression<String> {
-	
+@Since("2.2-dev33, INSERT VERSION (changers)")
+public class ExprPermissions extends PropertyExpression<Entity, String> {
+
 	static {
-		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] permissions (from|of) %players%", "[(all [[of] the]|the)] %players%'[s] permissions");
+		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY,
+				"[all [[of] the]|the] permissions (from|of) %entities%",
+				"[all [of] the] %entities%'[s] permissions"
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:57]:
```suggestion
				"%entities%'[s] permissions"
```
This should just be removed completely

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 14]
```diff
@@ -18,71 +18,135 @@
...
+	 * @return PermissionAttachment with Skript as the plugin
+	 */
+	private PermissionAttachment getPermission(Entity entity) {
+		Skript instance = Skript.getInstance();
+		for (PermissionAttachmentInfo info : entity.getEffectivePermissions()) {
+			PermissionAttachment attachment = info.getAttachment();
+			if (attachment == null)
+				continue;
+			if (attachment.getPlugin().equals(instance))
+				return attachment;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:147]:
```suggestion
			if (attachment != null && attachment.getPlugin().equals(instance))
				return attachment;
```

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 15]
```diff
@@ -18,71 +18,135 @@
...
+							attachment.unsetPermission(permission);
+						} else if (permissions.contains(permission)) {
+							attachment.unsetPermission(permission);
+						}
+					}
+				}
+			}
+			return;
+		}
+		for (Entity entity : getExpr().getAll(event)) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:112]: Is there a reason to be using getAll here?

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 16]
```diff
@@ -18,71 +18,135 @@
...
-	protected String[] get(Event e) {
-		final Set<String> permissions = new HashSet<>();
-		for (Player player : players.getArray(e))
-			for (final PermissionAttachmentInfo permission : player.getEffectivePermissions())
-				permissions.add(permission.getPermission());
-		return permissions.toArray(new String[permissions.size()]);
+	protected String[] get(Event event, Entity[] source) {
+		return getExpr().stream(event)
+				.flatMap(permissible -> permissible.getEffectivePermissions().stream())
+				.map(permission -> permission.getPermission())
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:72]: Can't this be a method reference?

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 17]
```diff
@@ -18,71 +18,135 @@
...
+				.toArray(String[]::new);
+	}
+
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE || mode == ChangeMode.DELETE)
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:78]: Replace with a switch statement?

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 18]
```diff
@@ -18,71 +18,135 @@
...
+					}
+				}
+			}
+			return;
+		}
+		for (Entity entity : getExpr().getAll(event)) {
+			PermissionAttachment attachment = getPermission(entity);
+			for (String permission : permissions)
+				attachment.setPermission(permission, true);
+		}
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:116]: This whole implementation is a mess and hard to follow. You should check for the `mode` in a switch statement and neatly put each mode's implementation in its own case.

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 19]
```diff
@@ -18,71 +18,135 @@
...
+
 	@Override
 	public boolean isSingle() {
 		return false;
 	}
-	
+
 	@Override
-	public Class<String> getReturnType() {
+	public Class<? extends String> getReturnType() {
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:125]: Why did you add this?

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java (hunk 19 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:128]: This is how Skript requires the getReturnType to be setup

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 20]
```diff
@@ -18,71 +18,146 @@
...
 	@Override
-	@Nullable
-	protected String[] get(Event e) {
-		final Set<String> permissions = new HashSet<>();
-		for (Player player : players.getArray(e))
-			for (final PermissionAttachmentInfo permission : player.getEffectivePermissions())
-				permissions.add(permission.getPermission());
-		return permissions.toArray(new String[permissions.size()]);
+	protected String[] get(Event event, Entity[] source) {
+		return getExpr().stream(event)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:70]: Isn't the `source` array the evaluated `getExpr()`? We don't want to evaluate this twice

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 21]
```diff
@@ -18,71 +18,146 @@
...
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case ADD:
+			case REMOVE:
+				return CollectionUtils.array(String[].class);
+			case DELETE:
+				return CollectionUtils.array();
+			default:
+				return null;
+		}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:86]:
```suggestion
		return switch (mode) {
			case ADD, REMOVE, DELETE -> CollectionUtils.array(String[].class);
			default -> null;
		};
```
We can simplify this now 🙂 
It's fine to just return the regular array for DELETE (no difference compared to an empty one)

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 22]
```diff
@@ -18,71 +18,146 @@
...
+			case ADD:
+				for (Entity entity : getExpr().getArray(event)) {
+					PermissionAttachment attachment = getPermission(entity);
+					for (String permission : permissions) {
+						attachment.setPermission(permission, true);
+					}
+				}
+				break;
+			default:
+				break;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:128]:
```suggestion
```
This can be removed since it doesn't do anything

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java (hunk 19 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprPermissions.java:138]: I mean I don't think anyone is going to be extending the class to where this change is necessary

File: src/main/java/ch/njol/skript/expressions/ExprPermissions.java [hunk 23]
```diff
@@ -18,71 +18,146 @@
...
+					for (PermissionAttachmentInfo info : entity.getEffectivePermissions()) {
+						PermissionAttachment attachment = info.getAttachment();
+						if (attachment == null)
+							continue;
+						for (String permission : attachment.getPermissions().keySet()) {
+							if (mode == ChangeMode.DELETE) {
+								attachment.unsetPermission(permission);
+							} else if (permissions.contains(permission)) {
+								attachment.unsetPermission(permission);
+							}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprPermissions.java:114]: Is there a reason to check that the permission is present? Can't we just do something like:
```
loop over attachments:
  loop over permissions to remove:
    remove permission from attachment
```
There doesn't seem to be any issue with removing a permission that is not present. It seems like we could make this more efficient

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
